### DISCUSSION TITLE
Prisma Dart client v4.0.0-alpha.0

### DISCUSSION THREAD

**Discussion by medz** [medz/dart-orm#292] [Announcements]
To install Prisma Dart client v4.0.0-alpha.0 run:

```sh
# If you are using Dart
dart pub add orm:4.0.0-alpha.0

# Or if you are using Flutter
flutter pub add orm:4.0.0-alpha.0
```

To upgrade to Prisma Dart client v4.0.0-alpha.0, Please follow the [migration guide](#migration-guide-from-v3) and update your `pubspec.yaml` file:

```yaml
dependencies:
  orm: 4.0.0-alpha.0
```

## What's Changed

- Whole project refactoring
- Remove any JSON serialization tool, now it's ready to use by just generating the client without any other dependencies and extra commands
- Switch from GraphQL protocol to JSON protocol
- Client takes a standalone Prisma engine instance
- Client and all input/output types are standalone and can be distributed to any Dart platform
- Add database field reference support
- Support `select` feature (incomplete, currently only support rough one-level Model fields)
- Support `include` feature (incomplete, currently only support rough one-level Model fields)
- `PrismaUnion` regression, now can structure nested inputs of multiple parameters via union
- `PrismaNull` regression, now support database `null` data setting
- DMMF, generator helpers regression, no need to depend on other packages, can directly use `orm` as the base package for developing Dart Prisma ecosystem packages
- Add `Decimal` type support (from [decimal package](https://pub.dev/package/decimal), exported by `orm` proxy)

### Step 1: Update your `pubspec.yaml` file

```yaml
dependencies:
  orm: ^4.0.0-alpha.0
```

### Step 2: Update prisma schame

Previous generator configuration:

```prisma
generator DartClient {
  provider = "dart run orm"
}
```

Current generator configuration:

```prisma
generator DartClient {
  provider = "dart run orm:client"
}
```

If you have `orm` installed globally, `provider` can use `prisma-client-dart` directly:

```prisma
generator DartClient {
  provider = "prisma-client-dart"
}
```

### Step 3: Configure your engine (in `schema.prisma`)

```prisma
generator CopyQueryEngine {
  provider = "dart run orm:binary"
  // If you have `orm` installed globally, `provider` can use `prisma-engine-binary` directly:
  // provider = "prisma-engine-binary"
}
```

> Currently, `orm` only supports Prisma binary engine, other engines will be supported in future updates.

### Step 4: Generate client

```sh
npx prisma generate
```

### Setp 5: Update your code

Previous import use `[path]/prisma_client.dart`, now use `[path]/client.dart`:

Now the client will generate three files:

- `client.dart` - Client implementation (interact with Prisma engine)
- `types.dart` - All input/output types
- `dmmf.dart` - DMMF data structure

```dart
import 'package:orm/orm.dart';
import 'package:orm/engines/binary.dart';

// Your generated client path.
import '[path]/client.dart';

final engine = BinaryEngine(
  datasource: 'db',
  schema: schema,
  url: Uri.parse('postgresql://seven@localhost:5432/prisma-dart'),
);
final prisma = PrismaClient(
  engine: engine,
);
```

**Comment by listepo** [comment]
I would like to help with testing, will this work on Flutter Mac?(sqlite version)

**Reply by medz** [reply]
Not yet, Flutter natively requires a brand new engine written in Rust, and I can’t estimate the arrival time yet. (A person’s time and power are ultimately limited)

**Reply by medz** [reply]
I'm waiting for https://github.com/prisma/prisma-engines/pull/4521

**Comment by FluffyDiscord** [comment]
Do I understand it correctly, that I can not use Prisma with SQLite natively on Android/iOS, but remote only?

**Reply by medz** [reply]
Currently, yes, it can only be used in the Server (binary) environment. I have not yet developed the Flutter engine for Android/iOS/macOS/Windows/Linux.

**Reply by medz** [reply]
I'm waiting for https://github.com/prisma/prisma-engines/pull/4521, with which I can continue to develop the Flutter native client, because now I use lower-level package development, once the Prisma engine update brings Change is disruptive. With it, Prisma dart can focus on its own implementation.
